### ISSUE TITLE
Map/Struct patterns should default to using same keys when updating

### ISSUE THREAD

**Issue by zachallaun** [zachallaun/mneme#41]
![image](https://user-images.githubusercontent.com/503938/233800484-d839dc26-6cba-419e-ba9c-32427a42786f.png)

In this example, the body has changed, but I've edited the assertion to only include the body key and not the headers. Mneme is only generating patterns for `%Req.Response{}` and `%Req.Response{body: "...", headers: [...]}`. In these cases, it would be nice to also generate a `%Req.Response{body: "..."}` and default to that.

This will require extracting some information (the map/struct keys) and passing it to `PatternBuilder` so that it can construct a pattern that most closely resembles the original assertion.

**Comment by zachallaun** [comment]
How should nested results be handled?

```elixir
{:ok, %Struct{foo: 1, bar: %{map_key: 2}}}
```

In the AST, structs will have an extra level of nesting than maps, but for the sake of this feature, I’d sort of want them treated at the same “level”. 

What if we extract the keys into a stack and the pattern builder can pull a set of keys off the stack to generate a pattern if it works for the value. 

```elixir
[[:foo, :bar], [:map_key]]
```

The pattern builder can pop the first matching off the stack and pass the remainder on to further calls. This would mean that every pattern builder would need to return both its pattern and modified context, though, which is maybe kinda annoying.

**Comment by zachallaun** [comment]
Fixed by cb29abeaa0561c8ac342e8cb751fb3cec2653f57.
